### PR TITLE
10168: Enable ignore_missing_imports only where it is needed.

### PR DESCRIPTION
**Pull request by wsanchez**
The current `mypy.ini` file sets `ignore_missing_imports=True` globally.

We should instead disable it specifically for packages that are known to lack type stubs. That ensures that for those where we expect type stubs, the Mypy error is not suppressed if they are missing.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10168
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: <github_username>, <github_usernames_if_more_authors>
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:<trac_ticket_number>, ticket:<another_if_more_in_one_PR>

Long description providing a summary of these changes.
(as long as you wish)
```

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -821,3 +819,213 @@ allow_incomplete_defs = True
 [mypy-twisted.words.protocols.jabber.jid]
 warn_return_any = False
 
+# Don't complain about imports of fake modules used in tests
+
+[mypy-__main__]
+ignore_missing_imports = True
+
+[mypy-automat]
```
Comment (adiroiban) [inline mypy.ini:827]: Is this only testing?

I see this in src/twisted/application/internet.py

```
from automat import MethodicalMachine
```

File: mypy.ini [hunk 2]
```diff
@@ -821,3 +819,213 @@ allow_incomplete_defs = True
...
+[mypy-twisted_renamed_helper]
+ignore_missing_imports = True
+
+[mypy-twisted_toptobottom_temp]
+ignore_missing_imports = True
+
+[mypy-uberpackage]
+ignore_missing_imports = True
+
+[mypy-utmp]
```
Comment (adiroiban) [inline mypy.ini:856]: Looks like this is used in `src/twisted/conch/unix.py` , outside of testing

**Review by adiroiban — APPROVED** [review]
I guess that this is ok.

I don't think that it make sense to manually double check if any of the dependencies do have mypy hints.

I guess that once one Twisted developer cares about typing and knows that for example h2
has typing, the list can be later updated.

Maybe @graingert  might want to double check... if he has time.

Thanks!

**Comment by graingert** [comment]
I've created https://github.com/twisted/twisted/pull/1584 which uses `# type: ignore[import]` which will  automatically remind us to remove the `ignore` annotation when stubs are made available

**Review by graingert — CHANGES_REQUESTED** [review]
I'm going to request changes because I think my solution in #1584 will be more convenient, because mypy will warn us when modules gain type hints/stubs, and "weird"/technical imports like `__main__`, `package.module`, `twisted_rebuild_fakelib` make more sense to be ignored locally than globally as they will never be stubbed and just clutter the mypy.init

https://github.com/twisted/twisted/pull/1584/files#diff-8f4abfb76eeb666cb2782ad79aee6d2bfaa9eaa0911b28edcd57f6e601f9c8d4R538

if you disagree feel free to dismiss my review

**Comment by graingert** [comment]
I'm closing this in favour of #1584 for now

**Pull request closed without merging**